src/paid_witnessing.rs

```diff
@@ -0,0 +1,270 @@
+use config;
+use error::Result;
+use graph;
+use mc_outputs;
+use rusqlite::Connection;
+use storage;
+
+pub fn calc_witness_earnings(
+    db: &Connection,
+    kind: &String,
+    from_main_chain_index: u32,
+    to_main_chain_index: u32,
+    address: &String,
+) -> Result<u32> {
+    let mut stmt = db.prepare_cached(
+        "SELECT COUNT(*) AS count FROM units \
+         WHERE is_on_main_chain=1 AND is_stable=1 \
+         AND main_chain_index>=? AND main_chain_index<=?",
+    )?;
+
+    let count = stmt.query_row(
+        &[
+            &to_main_chain_index,
+            &(to_main_chain_index + config::COUNT_MC_BALLS_FOR_PAID_WITNESSING + 1),
+        ],
+        |row| row.get(0),
+    ).unwrap_or(0);
+
+    ensure!(
+        count == config::COUNT_MC_BALLS_FOR_PAID_WITNESSING + 2,
+        "not enough stable MC units after to_main_chain_index"
+    );
+
+    mc_outputs::calc_earnings(
+        db,
+        kind,
+        from_main_chain_index,
+        to_main_chain_index,
+        address,
+    )
+}
+
+fn get_max_spendable_mci_for_last_ball_mci(last_ball_mci: u32) -> Option<u32> {
+    last_ball_mci.checked_sub(1 + config::COUNT_MC_BALLS_FOR_PAID_WITNESSING)
+}
+
+fn read_mc_unit_witnesses(db: &Connection, main_chain_index: u32) -> Result<Vec<String>> {
+    let mut stmt = db.prepare_cached(
+        "SELECT witness_list_unit, unit FROM units \
+         WHERE main_chain_index=? AND is_on_main_chain=1",
+    )?;
+
+    struct Row {
+        witness_list_unit: Option<String>,
+        unit: String,
+    }
+
+    let rows = stmt.query_map(&[&main_chain_index], |row| Row {
+        witness_list_unit: row.get(0),
+        unit: row.get(1),
+    })?;
+
+    let mut units = Vec::new();
+    for row in rows {
+        units.push(row?);
+    }
+
+    ensure!(units.len() == 1, "not 1 row on MC {}", main_chain_index);
+
+    let unit = if units[0].witness_list_unit.is_some() {
+        units[0].witness_list_unit.clone().unwrap().clone()
+    } else {
+        units[0].unit.clone()
+    };
+
+    storage::read_witness_list(db, &unit)
+}
+
+fn build_paid_witnesses(
+    db: &Connection,
+    unit_prop: graph::UnitProps,
+    witnesses: &Vec<String>,
+) -> Result<()> {
+    let main_chain_index = unit_prop.main_chain_index.unwrap_or(0);
+    let to_main_chain_index = main_chain_index + config::COUNT_MC_BALLS_FOR_PAID_WITNESSING;
+
+    let units = graph::read_descendant_units_by_authors_before_mc_index(
+        db,
+        &unit_prop,
+        &witnesses,
+        to_main_chain_index,
+    )?;
+
+    let unit = unit_prop.unit;
+    let unit_list = units
+        .iter()
+        .map(|s| format!("'{}'", s))
+        .collect::<Vec<_>>()
+        .join(", ");
+
+    let witness_list = witnesses
+        .iter()
+        .map(|s| format!("'{}'", s))
+        .collect::<Vec<_>>()
+        .join(", ");
+
+    if units.len() > 0 {
```

Comment:
how about when the units.len() == 0? when this happened, INSERT INTO paid_witness_events_tmp should still be executed.